### PR TITLE
Call _refresh_token only when session is alive

### LINKED ISSUE #4021: pydeephaven throws an exception about the failure to refresh auth token after the session is closed.
Found by @jnumainville when testing the Python client against a DnD worker.
```
Exception in thread Thread-13:
Traceback (most recent call last):
  File "/opt/homebrew/Cellar/python@3.9/3.9.17/Frameworks/Python.framework/Versions/3.9/lib/python3.9/threading.py", line 980, in _bootstrap_inner
    self.run()
  File "/Users/josephnumainville/Documents/deephaven-ipywidgets/venv/lib/python3.9/site-packages/deephaven_enterprise/client/controller.py", line 56, in run
    self.controller_client._process_one_event(x)
  File "/Users/josephnumainville/Documents/deephaven-ipywidgets/venv/lib/python3.9/site-packages/deephaven_enterprise/client/controller.py", line 394, in _process_one_event
    del self.query_map[serial]
KeyError: 1686761545175000000
Exception in thread Thread-9:
Traceback (most recent call last):
  File "/Users/josephnumainville/Documents/deephaven-ipywidgets/venv/lib/python3.9/site-packages/pydeephaven/session.py", line 317, in _refresh_token
    self._flight_client.authenticate(self._auth_handler)
  File "pyarrow/_flight.pyx", line 1387, in pyarrow._flight.FlightClient.authenticate
  File "pyarrow/_flight.pyx", line 78, in pyarrow._flight.check_flight_status
  File "pyarrow/error.pxi", line 100, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: FlightClient is closed

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/homebrew/Cellar/python@3.9/3.9.17/Frameworks/Python.framework/Versions/3.9/lib/python3.9/threading.py", line 980, in _bootstrap_inner
    self.run()
  File "/opt/homebrew/Cellar/python@3.9/3.9.17/Frameworks/Python.framework/Versions/3.9/lib/python3.9/threading.py", line 1306, in run
    self.function(*self.args, **self.kwargs)
  File "/Users/josephnumainville/Documents/deephaven-ipywidgets/venv/lib/python3.9/site-packages/pydeephaven/session.py", line 309, in _keep_alive
    self._refresh_token()
  File "/Users/josephnumainville/Documents/deephaven-ipywidgets/venv/lib/python3.9/site-packages/pydeephaven/session.py", line 320, in _refresh_token
    raise DHError("failed to refresh auth token") from e
pydeephaven.dherror.DHError: failed to refresh auth token
Exception in thread Thread-14:
Traceback (most recent call last):
  File "/Users/josephnumainville/Documents/deephaven-ipywidgets/venv/lib/python3.9/site-packages/pydeephaven/session.py", line 317, in _refresh_token
    self._flight_client.authenticate(self._auth_handler)
  File "pyarrow/_flight.pyx", line 1387, in pyarrow._flight.FlightClient.authenticate
  File "pyarrow/_flight.pyx", line 78, in pyarrow._flight.check_flight_status
  File "pyarrow/error.pxi", line 100, in pyarrow.lib.check_status
pyarrow.lib.ArrowInvalid: FlightClient is closed

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/homebrew/Cellar/python@3.9/3.9.17/Frameworks/Python.framework/Versions/3.9/lib/python3.9/threading.py", line 980, in _bootstrap_inner
    self.run()
  File "/opt/homebrew/Cellar/python@3.9/3.9.17/Frameworks/Python.framework/Versions/3.9/lib/python3.9/threading.py", line 1306, in run
    self.function(*self.args, **self.kwargs)
  File "/Users/josephnumainville/Documents/deephaven-ipywidgets/venv/lib/python3.9/site-packages/pydeephaven/session.py", line 309, in _keep_alive
    self._refresh_token()
  File "/Users/josephnumainville/Documents/deephaven-ipywidgets/venv/lib/python3.9/site-packages/pydeephaven/session.py", line 320, in _refresh_token
    raise DHError("failed to refresh auth token") from e
pydeephaven.dherror.DHError: failed to refresh auth token
```

### PR DESCRIPTION
**Pull request by jmao-denver**
Fixes #4021
Fixes DH-15110

### CODE REVIEW

File: py/client/pydeephaven/session.py [hunk 1]
```diff
@@ -330,7 +333,7 @@ def is_alive(self) -> bool:
                 return True
 
             try:
-                self.session_service.refresh_token()
+                self._flight_client.authenticate(self._auth_handler)
```
Comment (cpwright) [inline py/client/pydeephaven/session.py:336]: I don't understand the logic behind this line change.

File: py/client/tests/test_session.py [hunk 2]
```diff
@@ -41,6 +41,7 @@ def test_never_timeout(self):
             token2 = session._auth_token
             self.assertNotEqual(token1, token2)
         session.close()
+        sleep(400)
```
Comment (cpwright) [inline py/client/tests/test_session.py:44]: I assume we skip this because it is going to be very slow and annoy everyone?

**Review by cpwright — APPROVED** [review]

File: py/client/tests/test_session.py (hunk 2 above)
Comment (jmao-denver) [reply py/client/tests/test_session.py:44]: yes, until we allow the client to set the timeout on the session.

File: py/client/pydeephaven/session.py (hunk 1 above)
Comment (jmao-denver) [reply py/client/pydeephaven/session.py:336]: It was an overlooked bug. `self.session_service` doesn't have the `refresh_token` method anymore after the switch to flight based auth.

**Review by jnumainville — APPROVED** [review]

**Review by chipkent — APPROVED** [review]

**Pull request merged**